share/man/man8/obpsym.8
'\" te
.\"  Copyright 2001 Sun Microsystems, Inc.  All Rights Reserved
.\" The contents of this file are subject to the terms of the Common Development and Distribution License (the "License").  You may not use this file except in compliance with the License.
.\" You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE or http://www.opensolaris.org/os/licensing.  See the License for the specific language governing permissions and limitations under the License.
.\" When distributing Covered Code, include this CDDL HEADER in each file and include the License file at usr/src/OPENSOLARIS.LICENSE.  If applicable, add the following below this CDDL HEADER, with the fields enclosed by brackets "[]" replaced with your own identifying information: Portions Copyright [yyyy] [name of copyright owner]
.TH OBPSYM 8 "Dec 13, 2001"
.SH NAME
obpsym \- Kernel Symbolic Debugging for OpenBoot Firmware
.SH SYNOPSIS
.LP
.nf
\fBmodload\fR \fB-p\fR misc/obpsym
.fi

.SH DESCRIPTION
.sp
.LP
\fBobpsym\fR is a kernel module that installs OpenBoot callback handlers that
provide kernel symbol information to OpenBoot. OpenBoot firmware user interface
commands use the callbacks to convert numeric \fIaddresses\fR to kernel symbol
names for display purposes, and to convert kernel symbol names to numeric
\fIliterals\fR allowing symbolic names to be used as input arguments to user
interface commands.
.sp
.LP
Once \fBobpsym\fR is installed, kernel symbolic names may be used anywhere at
the OpenBoot firmware's user interface command prompt in place of a literal
(numeric) string. For example, if \fBobpsym\fR is installed, the OpenBoot
firmware commands \fBctrace\fR and \fBdis\fR typically display symbolic names
and offsets in the form \fImodname:symbolname + offset.\fR User interface
Commands such as \fBdis\fR can be given a kernel symbolic name such as
\fBufs:ufs_mount\fR instead of a numeric address.
.sp
.LP
Placing the command
.sp
.LP
\fBforceload: misc/obpsym\fR
.sp
.LP
into the \fBsystem\fR(4) file forces the kernel module \fBmisc/obpsym\fR to be
loaded and activates the kernel callbacks during the kernel startup sequence.
.sp
.LP
\fBobpsym\fR may be useful as a kernel debugger in situations where other
kernel debuggers are not useful. For example, on SPARC machines, if
\fBobpsym\fR is loaded, you may be able to use the OpenBoot firmware's
\fBctrace\fR command to display symbolic names in the stack backtrace after a
watchdog reset.
.SS "Kernel Symbolic Name Syntax"
.sp
.LP
The syntax for a kernel symbolic name is:
.sp
.LP
\fB\fR [ \fImodule-name\fR \fB:\fR ] \fIsymbol-name\fR
.sp
.LP
Where \fImodule-name\fR is the name of the kernel module that the symbol
\fIsymbol-name\fR appears in. A \fINULL\fR module name is taken as "all
modules, in no particular order" by \fBobpsym\fR. The module name \fBunix\fR is
equivalent to a \fINULL\fR module name, so that conflicts with words defined in
the firmware's vocabulary can be avoided.
.sp
.LP
Typically, OpenBoot firmware reads a word from the input stream and looks the
word up in its internal \fIvocabulary\fR before checking if the word is a
\fIliteral\fR. Thus, kernel symbols, such as \fBreset\fR may be given as
\fBunix:reset\fR to avoid the unexpected side effect of the firmware finding
and executing a matching word in its vocabulary.
.SH FILES
.sp
.ne 2
.na
\fB\fB/etc/system\fR\fR
.ad
.sp .6
.RS 4n
system configuration information file
.RE

.sp
.ne 2
.na
\fB\fB/platform/\fR\fIplatform-name\fR\fB/kernel/misc/obpsym\fR\fR
.ad
.sp .6
.RS 4n

.RE

.SH SEE ALSO
.sp
.LP
\fBkadb\fR(8), \fBkernel\fR(8), \fBmodload\fR(8), \fBmodunload\fR(8),
\fBuname\fR(1), \fBsystem\fR(4), \fBattributes\fR(5)
.sp
.LP
\fI\fR
.SH WARNINGS
.sp
.LP
Some OpenBoot firmware user interface commands may use system resources
incompatibly with the way they are used by the Unix kernel. These commands and
the use of this feature as a kernel debugger may cause interactions that the
Unix kernel is not prepared to deal with. If this occurs, the Unix kernel
and/or the OpenBoot firmware user interface commands may react unpredictably
and may panic the system, or may hang or may cause other unpredictable results.
For these reasons, the use of this feature is only minimally supported and
recommended to be used only as a kernel debugger of "last resort".
.sp
.LP
If a breakpoint or watchpoint is triggered while the console frame buffer is
powered off, the system can crash and be left in a state from which it is
difficult to recover. If one of these is triggered while the monitor is powered
off, you will not be able to see the debugger output.
.SH NOTES
.sp
.LP
\fIplatform-name\fR can be found using the \fB-i\fR option of \fBuname\fR(1)
.sp
.LP
\fBobpsym\fR is supported only on architectures that support OpenBoot firmware.
.sp
.LP
On some systems, OpenBoot must be completely RAM resident so the \fBobpsym\fR
symbol callback support can be added to the firmware, if the firmware doesn't
include support for the symbol callbacks. On these systems, \fBobpsym\fR may
complain that it requires that "you must use ramforth to use this module".
.sp
.LP
See the  for details on how to use the \fIramforth\fR command, how to place the
command into \fInvramrc\fR, and how to set \fIuse-nvramrc?\fR to \fBtrue\fR. On
systems with version 1.x OpenBoot firmware, \fInvramrc\fR doesn't exist, and
the \fIramforth\fR command must be typed manually after each reset, in order to
use this module.
.sp
.LP
Once installed, the symbol table callbacks can be disabled by using the
following OpenBoot firmware command:
.sp
.LP
\fB0 0 set-symbol-lookup\fR


usr/src/uts/sparc/os/obpsym.c
/*
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License (the "License").
 * You may not use this file except in compliance with the License.
 *
 * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
 * or http://www.opensolaris.org/os/licensing.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */

/*
 * Copyright 2008 Sun Microsystems, Inc.  All rights reserved.
 * Use is subject to license terms.
 */

/*
 * This module supports callbacks from the firmware
 * such that address and name lookups can work and use kernel symbol names.
 * For example "ctrace" will provide symbolic names, if they are available.
 * Also, normal firmware name to address lookups should work, though be
 * careful with clashing kernel names, such as "startup" and "reset" which
 * may be the firmware names and *not* the kernel names.
 *
 * The module locks the symbol tables in memory, when it's installed,
 * and unlocks them when it is removed.  The module is loaded automatically
 * on cobp systems and is replaced by forthdebug on all other systems.
 *
 * This file contains the actual code the does the lookups, and interfaces
 * with the kernel kobj stuff.  The transfer of data and control to/from
 * the firmware is handled in prom-dependent code.
 */

#include <sys/types.h>
#include <sys/param.h>
#include <sys/systm.h>
#include <sys/debug.h>
#include <sys/errno.h>
#include <sys/modctl.h>
#include <sys/kobj.h>
#include <sys/promif.h>
#include <sys/ddi.h>
#include <sys/sunddi.h>
#include <sys/reboot.h>
#include <sys/callb.h>

int obpsym_debug = 0;
#define	DPRINTF(str)		if (obpsym_debug) prom_printf(str)
#define	DPRINTF1(str, a)	if (obpsym_debug) prom_printf(str, a);
#define	DPRINTF2(str, a, b)	if (obpsym_debug) prom_printf(str, a, b);
#define	DXPRINTF		if (obpsym_debug > 1) prom_printf

int obpheld = 1;		/* Prevents unloading when set */

/*
 * name_to_value - translate a string into an address
 *
 * The string may be one of two forms - 'symname' or 'modname:symname'.
 * (We also accept 'unix:symname' as a synonymn for 'symname'.)
 * The latter form causes a kobj_lookup() to occur only in the given module,
 * the former causes a kobj_getsymvalue() on the entire kernel symbol table.
 *
 * mod_lock locking is not needed for the &modules list because
 * modctl structures are never unlinked from the &modules list.
 */
int
name_to_value(char *name, uintptr_t *value)
{
	register char *symname = name;
	register char *modname = "";
	char *p;
	int retval = 0;
	uintptr_t symvalue = 0;
	char c = '\0';

	/*
	 * we take names of the form: "modname:symbol", "unix:symbol", "symbol"
	 */
	if ((p = strchr(name, ':')) != NULL && p[1] != '\0') {
		symname = p + 1;
		modname = name;
		c = *p;
		*p = '\0';
	}

	if (*modname == '\0') {
		symvalue = kobj_getsymvalue(symname, 0);
	} else  {
		struct modctl *mp = &modules;

		do {
			if (strcmp(modname, mp->mod_modname) == 0) {
				symvalue = kobj_lookup(mp->mod_mp, symname);
				break;
			}
		} while ((mp = mp->mod_next) != &modules);
	}

	if (symvalue == 0)
		retval = -1;
	if (c != '\0')		/* Restore incoming cstr */
		*p = c;

	*value = symvalue;
	return (retval);
}

/*
 * value_to_name - translate an address into a string + offset
 *
 * mod_lock locking is not needed fro the modules list because
 * modctl structures are never unlinked from the &modules list.
 */
ulong_t
value_to_name(uintptr_t value, char *symbol)
{
	struct modctl *modp = &modules;
	ulong_t offset;
	char *name;

	DPRINTF1("value_to_name: Looking for %p\n", (void *)value);

	do {
		if (modp->mod_mp &&
		    (name = kobj_searchsym(modp->mod_mp, value, &offset))) {
			(void) strcpy(symbol, modp->mod_modname);
			(void) strcat(symbol, ":");
			(void) strcat(symbol, name);
			return (offset);
		}
	} while ((modp = modp->mod_next) != &modules);

	*symbol = '\0';
	return ((ulong_t)-1l);
}

/*
 * loadable module wrapper
 */
static struct modlmisc modlmisc = {
	&mod_miscops, "OBP symbol callbacks"
};

static struct modlinkage modlinkage = {
	MODREV_1, (void *)&modlmisc, NULL
};

/*ARGSUSED*/
static boolean_t
reset_callbacks(void *arg, int code)
{
	extern void set_sym_callbacks();

	if (code == CB_CODE_CPR_RESUME)
		set_sym_callbacks();
	return (B_TRUE);
}

int
_init(void)
{
	int retval;
	extern int install_callbacks(void);
	extern void remove_callbacks(void);

	if (install_callbacks() != 0)
		return (ENXIO);

	if (boothowto & RB_HALT)
		debug_enter("obpsym: halt flag (-h) is set.\n");

	retval = mod_install(&modlinkage);

	/*
	 * if load fails remove callback and unlock symbols
	 */
	if (retval) {
		printf("obpsym: Error %d installing OBP syms module\n", retval);
		remove_callbacks();
	}
	else
		(void) callb_add(reset_callbacks, 0, CB_CL_CPR_OBP, "obpsym");

	return (retval);
}

int
_fini(void)
{
	int retval;
	extern void remove_callbacks(void);

	if (obpheld != 0)
		return (EBUSY);

	retval = mod_remove(&modlinkage);

	/*
	 * if unload succeeds remove callback and unlock symbols
	 */
	if (retval == 0) {
		remove_callbacks();
	}
	return (retval);
}

int
_info(struct modinfo *modinfop)
{
	return (mod_info(&modlinkage, modinfop));
}
